[FEAT] 서비스에 보여줄 데이터를 쌓기

giantim opened the pull request:
- 카카오 스케쥴러, 인스타그램 스케쥴러가 기존 구조에서는 데이터를 수집하는 동안 데이터베이스와 연결이 끊어질 가능성이 있던 부분을 수정했습니다.
- 카카오 스케쥴러가 수집한 데이터를 저장하는 과정에서 Document를 변환할 때 x, y 좌표 값이 잘못 들어가던 문제를 데이터 수집중에 발견해서 수정했습니다.
- 서울 행정구 25곳을 데이터베이스에 저장해 놓았습니다. -> 실험 결과 25개 구 정보를 한번에 카카오 스케쥴러에 넣고 돌리면 429를 던져줘서 더 이상 데이터 수집을 못합니다. 20번째 구에 대한 가게 정보를 탐색하던 도중 429 요청이 넘어오고 더 이상 카카오에서 정보 수집을 하지 못하도록 막았습니다. 질문 게시판 확인 결과 카카오에서는 시간 당 요청수로 봇을 탐지한다고 합니다. 따라서 카카오 스케쥴러가 한번에 최대 10개 구 정도만 탐색하도록 변경해야 합니다.
- kakao-scheduler모듈에 PlaceCollect 이라는 테스트 코드로 데이터를 수집했습니다. approve되면은 해당 테스트 코드는 삭제하고 머지하도록 하겠습니다. 

- 인스타그램 스케쥴러는 현재 기존 데이터를 데이터베이스에서 모두 지운 후 새로 수집합니다. 그래서 instagram-scheduler모듈에 InstagramCollect 이라는 테스트 코드를 만들어서 데이터를 수집했습니다.
- 인스타그램 스케쥴러가 모든 가게 정보를 받아와서 크롤링을 시도하면 자꾸 중간에 에러가 발생해서 1만개 씩 가게 정보를 나눠서 데이터를 수집했습니다.
- 결과가 너무 부정확해서 로직 개선이 필요합니다. 

closed: #179

giantim committed:
refactor: 데이터 수집을 위한 코드 개선

- 트랜잭션 범위 조정
- 인스타그램 검색을 위한 PlaceNameParser 개선
- KakaoScheduler 버그 수정

giantim committed:
refactor: 서브모듈 버전 수정

moonyoungCHAE commented on hashtagmap-kakao-scheduler/src/main/java/com/songpapeople/hashtagmap/scheduler/domain/factory/PlaceFactory.java:
```diff
@@ -18,7 +18,11 @@
 
     public static Place from(Document document) {
-        Location location = new Location(new Point(document.getLatitude(), document.getLongitude()), document.getRoadAddressName());
+        /*
```
lat lon 정말 문제군요 ㅜㅜ

moonyoungCHAE commented on hashtagmap-instagram-scheduler/src/main/java/com/songpapeople/hashtagmap/scheduler/InstagramScheduler.java:
```diff
@@ -32,6 +31,11 @@ public void update() {
+
+    @Transactional
+    void saveCrawlingResult(List<CrawlingResult> crawlingResults) {
         for (CrawlingResult crawlingResult : crawlingResults) {
```
deleteAll을 여기다가 해줘야 할 것 같아요!
deleteAll하고 save하다가 오류 생기면 롤백하는 형식으로

giantim replied:
맞네요!
그럼 deleteAll도 다른 메서드로 분리해서 트랜잭션 범위를 나눠서 가져가는게 좋아보이네요 :)

moonyoungCHAE replied:
엇 이 부분 하나의 transaction 안에서 delete + save 하자는 말이었어요!!
1. findAll
2. 크롤링
3. 메서드 (deleteAll + saveAll)

저렇게 하면 크롤링하다가 문제 생기면 지운 게 복구 안 되지 않나요?!

moonyoungCHAE commented on hashtagmap-instagram-scheduler/src/test/java/com/songpapeople/hashtagmap/collect/InstagramCollect.java:
```diff
@@ -0,0 +1,73 @@
+
+@ActiveProfiles("collect")
+@SpringBootTest
+public class InstagramCollect {
```
```suggestion
public class InstagramCollectTest {
```

moonyoungCHAE commented on hashtagmap-instagram-scheduler/src/test/java/com/songpapeople/hashtagmap/collect/InstagramCollect.java:
```diff
@@ -0,0 +1,73 @@
+            여기는 테스트 코드에서 findAll()해온 place를 나눠서 진행한게 한번에 넣고 테스트를 돌리면
+            가끔씩 에러가 나면서 멈추는 경우가 생겨서 여러 단계로 나눠서 데이터를 수집했습니다.
+         */
+        List<Place> places = placeRepository.findAll();
```
여기 ParameterizedTest로 했으면 조금 덜 귀찮았을 것 같아요! 어차피 지울거니까 하자는 건 아니구요~

giantim replied:
데이터 수집 테스트 코드들이 findAll한 엔티티들을 범위를 나눠서 수동으로 실행한 이유가 요청이 너무 많아지면 429가 발생하더라고요!
인스타 뿐만 아니라 카카오 api도 429를 내려주더라고요. 그래서 직접 값을 나눠가면서 시간을 두고 수동으로 실행했습니다 🙃

moonyoungCHAE replied:
아하 시간 때문에 그랬군요

moonyoungCHAE commented on hashtagmap-kakao-scheduler/src/main/java/com/songpapeople/hashtagmap/scheduler/domain/KakaoSchedulerTask.java:
```diff
@@ -32,27 +31,30 @@ public KakaoSchedulerTask(ZoneRepository zoneRepository, PlaceRepository placeRe
     }
+
+    @Transactional
+    void savePlaces(List<Place> places) {
```
개인적으로 궁금해서 실험해봤는데
saveAll이나 updateAndInsert에서 쓰인 batch성 쿼리에서는 transactional을 안 붙여도 알아서 rollback되네요~~ 공유

giantim replied:
saveAll도 내부에 그런 기능이 있는지 몰랐네요~ 역시 호기심천국 👍 👍 👍 👍 👍

pci2676 replied:
그래도 붙여주는게 좋을 것 같아요
다른 사람이 그 정보를 모르면 롤백이 안된다고 생각하고 그게 전략이라고 오해할 수 있을 것 같아요~

moonyoungCHAE commented on hashtagmap-kakao-scheduler/src/test/java/com/songpapeople/hashtagmap/collect/PlaceCollect.java:
```diff
@@ -0,0 +1,93 @@
+
+@ActiveProfiles("collect")
+@SpringBootTest
+public class PlaceCollect {
```
```suggestion
public class PlaceCollectTest {
```

giantim replied:
데이터 수집한 테스트 코드는 머지할 때 지우고 머지하려고 하는데 어떻게 생각하세요? :)

ordinCode replied:
음...나도 지우는게 맞는거 같은데 혹시 모를 상황이 있으니까 그냥 놔두는게 나을 것 같기도하네요
차피 active를 collect로 해야 도는거니까

그리고 최대한 빨리 수집로직으로 개선해서 그때지우는 걸로..?

moonyoungCHAE replied:
전 그냥 지워도 될 거 같아요~
이번에 데이터 수집이 일단은 끝났기 때문에!

giantim replied:
그럼 이번에 머지할 때 빼고 올리고 제 로컬에는 코드를 남겨둘게요!
나중에 다른 분들이 데이터를 수집하고 싶다 할 때 코드를 공유할 수 있도록 하겠습니다 :)

moonyoungCHAE commented on hashtagmap-kakao-scheduler/src/test/java/com/songpapeople/hashtagmap/collect/PlaceCollect.java:
```diff
@@ -0,0 +1,93 @@
+    void customCollectPlace() {
+        List<Zone> zones = zoneRepository.findByActivated();
+        List<Rect> rects = RectFactory.from(zones);
+        List<Rect> otherRects = new ArrayList<>(rects.subList(19, 25));
```
이 메서드는 조금 이해가 안됩니당,, rects와 otherRects의 의미는 뭔가요?

giantim replied:
이거는 25개 구를 한번에 가게 데이터 수집 요청을 보내면 429가 나서 수동으로 잘라서 넣어줬어요 :)
위에 써놓았습니다!

moonyoungCHAE reviewed (commented):
라빈! 이런 저런 실험하느라 고생했습니당!!! 👍 덕분에 배포할 때 데이터 따악 생겼네요
그리고 429의 정체는 카카오였다는 거죠?
그러면 key를 여러 개 파서 proxy처럼 돌려끼는 방법도 있겠네요.

ordinCode commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/util/PlaceNameParser.java:
```diff
@@ -1,12 +1,14 @@
+            PlaceNameType placeNameType = optionalPlaceNameType.get();
+            return placeNameType.parsePlaceName(parsedPlaceName);
         }
         return parsedPlaceName;
```
```suggestion
        Optional<PlaceNameType> optionalPlaceNameType = PlaceNameType.find(parsedPlaceName);
        return optionalPlaceNameType
                .map(placeNameType -> placeNameType.parsePlaceName(parsedPlaceName))
                .orElseGet(()->parsedPlaceName);
```

isPresent 말고 이렇게 하면 더 좋을것 같아요!

giantim replied:
굿 :)

ordinCode commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/util/PlaceNameType.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.function.Function;
+
+public enum PlaceNameType {
+    ORIGINAL_BRANCH("본점",
+            placeName -> placeName.substring(0, placeName.length() - 2)),
+    DIRECT_BRANCH("직영점",
+            placeName -> placeName.substring(0, placeName.length() - 3)),
+    CHINESE_BRANCH("반점",
+            placeName -> placeName),
+    BRANCH("점",
+            placeName -> placeName.substring(0, placeName.length() - 1));
+
```
이넘 굿!

ordinCode reviewed (approved):
라빈 데이터수직하느라 고생많았습니다! 큰일을 해냈네요.

- PlaceNameParserTest에 본점 반점 직영점 케이스에 대해서도 테스트 추가하면 좋을 것 같아요.

pci2676 commented on hashtagmap-instagram-scheduler/src/main/java/com/songpapeople/hashtagmap/scheduler/InstagramScheduler.java:
```diff
@@ -22,7 +22,6 @@
     private final InstagramRepository instagramRepository;
     private final InstagramScheduleService instagramScheduleService;
 
-    @Transactional
```
`@Transactional` 땐 이유가 뭔가요??
`instagramPostRepository.deleteAll();`가 실행된 이후에 에러가 터지면 `deleteAll()` 했던게 롤백이 안될것 같은데 의도된건가요?
`@Transactional` 붙이고 테스트 돌려보니까 정상적으로 돌아가는것으로 보이거든요!

giantim replied:
이 부분 제 실수가 있었네요 :)
지적 감사합니다!

pci2676 commented on hashtagmap-kakao-scheduler/src/main/java/com/songpapeople/hashtagmap/scheduler/domain/factory/PlaceFactory.java:
```diff
@@ -18,7 +18,11 @@
+        /*
+            여기가 변경된 이유는 x, y 좌표를 반대로 넣고 있어서 오류가 발생했었습니다.
+         */
+        Location location = new Location(new Point(document.getLongitude(), document.getLatitude()),
```
`point` 도메인에서는 `public Point(final String latitude, final String longitude)` 이렇게 받고 있는데 이게 무슨일인거죠..ㅠㅠ
`document`에서 들고있는게 반대라는 건가요? 그렇다면 메서드를 바꿔줘야 할 것 같아요!

giantim replied:
이 부분이 정말 헷갈려서 내일 다시 꼼꼼히 짚어보고 수정해서 리뷰 보내도록 해볼께요 😥
사실 지금 저도 너무 헷갈려요!

pci2676 reviewed (changes requested):
애증의 위도 경도...ㅠㅠ
고생하셨습니다!
질문남겼으니 답변 부탁드립니다!

DunDung commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/util/PlaceNameType.java:
```diff
@@ -0,0 +1,34 @@
+            placeName -> placeName.substring(0, placeName.length() - 2)),
+    DIRECT_BRANCH("직영점",
+            placeName -> placeName.substring(0, placeName.length() - 3)),
+    CHINESE_BRANCH("반점",
```
네이밍 센스 굿

DunDung reviewed (approved):
라빈 고생하셨슴다:)

giantim committed:
refactor: 리뷰 반영

- PlaceNameParser 로직 수정 및 테스트 코드 추가
- latitude, longitude를 api 규격에 맞도록 수정
- 인스타그램 스케쥴러 동작 방식 수정

giantim committed:
refactor: 데이터 수집 코드 제거
